proposals/multi-field-value-classes.md

```diff
@@ -0,0 +1,601 @@
+
+The reasons why the feature is implemented in Kotlin for pre-Valhalla JVMs too are:
+
+* Valhalla is only JVM, while Kotlin is not.
+* Valhalla would hardly be adopted in Android in any nearest future because Android Runtime is still compatible with JVM 1.8.
+* Adoption of new JVM’s is quite slow.
+
+
+As any other class, value classes can declare member (except inner classes), be a member of some other class, have generics, extensions and other language features.
+
+MFVC are currently only supported in Kotlin/JVM so description below is related only to it.
+
+## Current limitations
+
+Limitations for MFVC are similar to the [limitations of inline classes](https://github.com/Kotlin/KEEP/blob/master/proposals/inline-classes.md#current-limitations) but primary constructors can have arbitrary number of parameters greater than 1 (case of 1 parameter is inline class).
+
+## Java interoperability
+
+Each variable of MFVC is stored as several variables corresponding to their state. The same is for fields. if some MFVC contains other MFVC fields, they are flattened recursively. It means that the actual representation of variable (field)  of `DSegment` type (declared above) in the code is 4 `Double` variables (fields) corresponding to the both coordinates of its points.
+
+## Boxing
+
+Each multi-field value class has its own wrapper that is represented as a usual class on JVM. This wrapper is needed to box values of value class types and use it where it's impossible to use unboxed values.
+
+**Rules for boxing** are the following:
+
+* **Boxed when used as other type.** It means that if one uses MFVC as nullable type, implemented interface or `Any`, the boxing will happen: `dRectangle.area` doesn’t require the boxing while `(dRectangle as FigureWithArea<Double>).area` does.
+* **When used as type parameter (including function types).** It happens because type parameters are erased in JVM. It is the same as for inline classes. However, there are exception for the rule:
+    * When type parameter upper bound is MFVC.
+    * When type parameter is a type parameter of inline function or lambda.
+* **Returning from not inline function.** It happens because it is impossible to pass several objects between frames in JVM. The exception is chain of MFVC getters: the compiler replaces `wrapper.segment.p1.x` with getter `wrapper.`getSegment-p1-x`()` instead of `wrapper.getSegment().getP1().getX()`. And this complex getter escapes boxing when getter implementation was initially just reading the field.
```

Comment:
This should allow backticks to be properly rendered in inline code block.
```suggestion
* **Returning from not inline function.** It happens because it is impossible to pass several objects between frames in JVM. The exception is chain of MFVC getters: the compiler replaces `wrapper.segment.p1.x` with getter ``wrapper.`getSegment-p1-x`()`` instead of `wrapper.getSegment().getP1().getX()`. And this complex getter escapes boxing when getter implementation was initially just reading the field.
```